### PR TITLE
implement PUT route

### PR DESCRIPTION
**Pull request by mmcode9407**

### CODE REVIEW

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task goal
The goal of the task was to implement an endpoint for updating existing trains. The endpoint should update the `trains.json` file with the updated train data.

### Requirements
The requirements specify that the `trains.json` file should be updated with the updated train data. The payload object should contain the following fields: `id`, `trainExpressName`, `countryOfOrigin`, `yearOfConstruction`, `maxKilometerPerHour`, `destinationFrom`, `destinationTo`. The required endpoint schema is `PUT /trains/:id`.

### Code changes
The code changes include the addition of a new endpoint `PUT /trains/:id` in the `index.js` file. The endpoint reads the `trains.json` file, finds the train with the given id, updates the specified fields with the new data, and then writes the updated data back to the `trains.json` file.

### Review
The code changes cover the task requirements. The endpoint schema matches the required schema. The code reads from and writes to the `trains.json` file as required. The code correctly updates the specified fields in the train data.

However, there is a potential issue with the current implementation. If a field in the payload object is set to a falsy value (such as 0 or an empty string), it will not be updated due to the use of the logical OR operator (`||`). This could be a problem if, for example, we want to update a field to an empty string or zero.

To fix this issue, you could check if the field exists in the payload object instead of checking its truthiness:

```diff
- existingDataItem[field] = newData[field] || existingDataItem[field];
+ existingDataItem[field] = newData.hasOwnProperty(field) ? newData[field] : existingDataItem[field];
```

This will correctly update the field even if its new value is falsy.

**Review by NerdbordMentor — APPROVED** [review]
## Code Review Report

### Task goal
Implement an endpoint for updating existing trains.

### Requirements
- `trains.json` should be updated with updated train data.

### Code changes
The provided code changes are in the file `index.js`. The changes include the implementation of a new endpoint `PUT /trains/:id` which is used to update the data of an existing train.

### Review

The code changes seem to cover the task requirements. The `PUT /trains/:id` endpoint has been implemented. The endpoint reads the existing data from `trains.json`, finds the train with the given id, updates the fields of the train with the new data provided in the request body, and then writes the updated data back to `trains.json`.

The code also handles error scenarios such as when there is an error reading or writing to `trains.json`, and when a train with the given id is not found. In these cases, appropriate HTTP status codes and error messages are sent in the response.

The fields that can be updated are specified in the `fieldsToUpdate` array. This includes all the fields mentioned in the example payload object in the task requirements.

After updating the train data, the updated train object is sent in the response. This is not explicitly mentioned in the task requirements, but it is a common practice in REST APIs.

In conclusion, the code changes seem to fulfill the task requirements. However, without knowing more about the overall application and its context, it's hard to say if there might be other considerations or potential issues. For example, there might be a need for validation of the new data, or for concurrency control if multiple requests could be updating the same train at the same time.
